Tensile/Components/LocalRead.py

```diff
@@ -218,32 +212,76 @@ def __call__(self, writer, bufferIdx, iui, epsi, tP):
+                              divisorName = tP["lsc"]
+                          else:
+                            if tP["grcv"]:
+                              divisorName = tP["lsp"]
+                            else:
+                              divisorName = tP["lvp"]
+                          divisor = kernel[divisorName]
+                          if divisor < writer.kernel["WavefrontSize"]:
+                            # DirectToLds + above conditions, swap offset_val bits to adjust LDS offset
+                            # need to swap col and row index
+                            #waveDiff = 1 if kernel["WaveSeparateGlobalRead%c"%tc] else kernel["NumThreads"]//kernel["WavefrontSize"]
+                            waveDiff = 1
+                            if (not kernel["WaveSeparateGlobalRead%c"%tc]) and tP["glvw"] == 1:
+                              waveDiff = kernel["NumThreads"]//kernel["WavefrontSize"]
+                            scale = tP["nrc"]
+                            scaleShift = int(log2(scale)) # assuming scale is power of 2
+                            scaleShift += int(log2(waveDiff))
+                            ldsLineSize = kernel["WavefrontSize"] * kernel["GlobalLoadVectorWidth%c"%tc] * tP["bpe"]
+                            ldsLineSize //= scale
+                            maskBitsLow = (scale - 1) * ldsLineSize
+                            maskBitsHigh = maskBitsLow * scale * waveDiff
+                            maskBitsAll = (maskBitsLow | maskBitsHigh)
+                            tmp1 = offset_val & maskBitsLow
+                            tmp2 = offset_val & maskBitsHigh
+                            tmp1 <<= scaleShift
+                            tmp2 >>= scaleShift
+                            tmp1 = tmp1 | tmp2
+                            offset_val = (offset_val & (~maskBitsAll)) | tmp1
+
+                          # offset conversion for DirectToLds
+                          # TLU=0 case, modify bit3-6 of offset_val as follows
+                          # (bit2<<3) | (bit3 <<1) | (bit4>>2) | (bit5>>2)
+                          bit2 = offset_val & 4
+                          bit3 = offset_val & 8
+                          bit4 = offset_val & 16
+                          bit5 = offset_val & 32
+                          if (kernel["VectorWidth"] * tP["bpe"] == 8):
+                            # dword_x2 case
+                            # (bit2<<3) | (bit3 >>1) | (bit4>>1) | (bit5>>1)
+                            newVal = (bit2<<3) | (bit3 >>1) | (bit4>>1) | (bit5>>1)
+                          else:  #if (kernel["VectorWidth"] * tP["bpe"] == 16):  # most preferred case
+                            # dword_x4 case
+                            # (bit2<<3) | (bit3 <<1) | (bit4>>2) | (bit5>>2)
+                            newVal = (bit2<<3) | (bit3 <<1) | (bit4>>2) | (bit5>>2)
+                          offset_val = offset_val & (~0x3c)
+                          offset_val = offset_val | newVal
```

Comment:
Why bitswap instead of calculate based on other params like eIdx, rIdx as done above?

Reply:
Previous logic does not work with NN.
Further conversion is necessary at later stages.
Essentially, only the final offset matters to this conversion.
We do not need to care about intermediate result.
Converting address at final stage is the simplest way to cover both NN and NT.